File: DSA/Hard/C/solutions.c
/* DSA 100 - Hard | C solutions. Signatures are contest-style; malloc'd returns should be freed by caller. */
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <limits.h>
#include <stdbool.h>
#include <math.h>
typedef struct ListNode{int val;struct ListNode*next;}ListNode;typedef struct TreeNode{int val;struct TreeNode*left,*right;}TreeNode;
int q71_trap(int*h,int n){int l=0,r=n-1,lm=0,rm=0,a=0;while(l<=r){if(h[l]<=h[r]){lm=lm>h[l]?lm:h[l];a+=lm-h[l++];}else{rm=rm>h[r]?rm:h[r];a+=rm-h[r--];}}return a;}
int q72_first_missing_positive(int*a,int n){for(int i=0;i<n;i++)if(a[i]<1||a[i]>n)a[i]=n+1;for(int i=0;i<n;i++){int v=abs(a[i]);if(v<=n&&a[v-1]>0)a[v-1]=-a[v-1];}for(int i=0;i<n;i++)if(a[i]>0)return i+1;return n+1;}
double q73_median_two_sorted(int*A,int m,int*B,int n){if(m>n)return q73_median_two_sorted(B,n,A,m);int lo=0,hi=m,total=m+n;while(lo<=hi){int i=(lo+hi)/2,j=(total+1)/2-i;int al=i?A[i-1]:INT_MIN,ar=i<m?A[i]:INT_MAX,bl=j?B[j-1]:INT_MIN,br=j<n?B[j]:INT_MAX;if(al<=br&&bl<=ar)return total%2?(al>bl?al:bl):((double)(al>bl?al:bl)+(ar<br?ar:br))/2;if(al>br)hi=i-1;else lo=i+1;}return 0;}
int q74_largest_rectangle(int*h,int n){int*st=malloc((n+1)*sizeof(int));int top=0,ans=0;for(int i=0;i<=n;i++){int x=i==n?0:h[i];while(top&&h[st[top-1]]>x){int p=st[--top],left=top?st[top-1]:-1;int area=h[p]*(i-left-1);if(area>ans)ans=area;}st[top++]=i;}free(st);return ans;}
int q75_max_product(int*a,int n){int mx=a[0],mn=a[0],ans=a[0];for(int i=1;i<n;i++){if(a[i]<0){int t=mx;mx=mn;mn=t;}mx=mx*a[i]>a[i]?mx*a[i]:a[i];mn=mn*a[i]<a[i]?mn*a[i]:a[i];if(mx>ans)ans=mx;}return ans;}
long long q76_count_inversions(int*a,int n){int*tmp=malloc(n*sizeof(int));long long ans=0;void merge(int l,int r){if(r-l<2)return;int m=(l+r)/2;merge(l,m);merge(m,r);int i=l,j=m,k=l;while(i<m||j<r){if(j==r||(i<m&&a[i]<=a[j]))tmp[k++]=a[i++];else tmp[k++]=a[j++],ans+=m-i;}for(i=l;i<r;i++)a[i]=tmp[i];}merge(0,n);free(tmp);return ans;}
/* Q77 */
char*q77_min_window(const char*s,const char*t){int need[128]={0},have[128]={0};for(int i=0;t[i];i++)need[(unsigned char)t[i]]++;int missing=strlen(t),l=0,st=0,best=INT_MAX;for(int r=0;s[r];r++){unsigned char c=s[r];have[c]++;if(have[c]<=need[c])missing--;while(!missing){if(r-l+1<best)best=r-l+1,st=l;unsigned char d=s[l++];if(--have[d]<need[d])missing++;}}char*out=malloc((best==INT_MAX?0:best)+1);if(best==INT_MAX){out[0]=0;return out;}memcpy(out,s+st,best);out[best]=0;return out;}
/* Q78: word-break reconstruction; words[] has count w. Returns malloc'd array of chosen word pointers. */
char**q78_word_break(const char*s,const char**words,int w,int*outn){int n=strlen(s);int*pre=malloc((n+1)*sizeof(int));for(int i=0;i<=n;i++)pre[i]=-1;pre[0]=0;for(int i=0;i<n;i++)if(pre[i]>=0)for(int z=0;z<w;z++){int len=strlen(words[z]);if(i+len<=n&&pre[i+len]<0&&!strncmp(s+i,words[z],len))pre[i+len]=z+1;}if(pre[n]<0){*outn=0;free(pre);return NULL;}char**r=malloc((n+1)*sizeof(char*));int c=0,pos=n;while(pos){int z=pre[pos]-1,len=strlen(words[z]);r[c++]=(char*)words[z];pos-=len;}for(int i=0;i<c/2;i++){char*t=r[i];r[i]=r[c-1-i];r[c-1-i]=t;}*outn=c;free(pre);return r;}
/* Q79 */
bool q79_regex_match(const char*s,const char*p){
 int n=strlen(s),m=strlen(p); char*dp=calloc((n+1)*(m+1),1);
 #define D(i,j) dp[(i)*(m+1)+(j)]
 D(0,0)=1;
 for(int j=2;j<=m;j++) if(p[j-1]=='*') D(0,j)=D(0,j-2);
 for(int i=1;i<=n;i++) for(int j=1;j<=m;j++) {
  if(p[j-1]=='*') D(i,j)=D(i,j-2)||((p[j-2]=='.'||p[j-2]==s[i-1])&&D(i-1,j));
  else D(i,j)=(p[j-1]=='.'||p[j-1]==s[i-1])&&D(i-1,j-1);
 }
 bool ans=D(n,m); free(dp); return ans;
}
int q80_edit_distance(const char*a,const char*b){int n=strlen(b);int*dp=malloc((n+1)*sizeof(int));for(int j=0;j<=n;j++)dp[j]=j;for(int i=1;a[i-1];i++){int prev=dp[0];dp[0]=i;for(int j=1;j<=n;j++){int old=dp[j],v=prev+(a[i-1]!=b[j-1]);int x=dp[j]+1,y=dp[j-1]+1;if(x<v)v=x;if(y<v)v=y;dp[j]=v;prev=old;}}int ans=dp[n];free(dp);return ans;}
int q81_longest_pal_subseq(const char*s){int n=strlen(s);int*dp=calloc(n,sizeof(int));for(int i=n-1;i>=0;i--){int prev=0;dp[i]=1;for(int j=i+1;j<n;j++){int old=dp[j];dp[j]=s[i]==s[j]?prev+2:(dp[j]>dp[j-1]?dp[j]:dp[j-1]);prev=old;}}int ans=n?dp[n-1]:0;free(dp);return ans;}
ListNode*q82_merge_k_lists(ListNode**a,int k){ListNode d={0},*t=&d;while(1){int bi=-1;for(int i=0;i<k;i++)if(a[i]&&(bi<0||a[i]->val<a[bi]->val))bi=i;if(bi<0)break;t->next=a[bi];t=t->next;a[bi]=a[bi]->next;}return d.next;}
ListNode*q83_reverse_k_group(ListNode*h,int k){ListNode*d=malloc(sizeof(ListNode));d->next=h;ListNode*g=d;while(1){ListNode*p=g;for(int i=0;i<k&&p;i++)p=p->next;if(!p)break;ListNode*cur=g->next,*prev=p->next;for(int i=0;i<k;i++){ListNode*n=cur->next;cur->next=prev;prev=cur;cur=n;}ListNode*old=g->next;g->next=prev;g=old;}ListNode*r=d->next;free(d);return r;}
/* Q84 */
typedef struct CacheNode{int key,val;struct CacheNode*prev,*next;}CacheNode;typedef struct{int cap,size;CacheNode*head,*tail;}LRU;static void unlinkNode(LRU*c,CacheNode*x){x->prev->next=x->next;x->next->prev=x->prev;}static void frontNode(LRU*c,CacheNode*x){x->next=c->head->next;x->prev=c->head;c->head->next->prev=x;c->head->next=x;}/* Hash-map omitted: use linear key scan for dependency-free C. */int q84_get(LRU*c,int key){for(CacheNode*x=c->head->next;x!=c->tail;x=x->next)if(x->key==key){unlinkNode(c,x);frontNode(c,x);return x->val;}return -1;}void q84_put(LRU*c,int key,int val){for(CacheNode*x=c->head->next;x!=c->tail;x=x->next)if(x->key==key){x->val=val;unlinkNode(c,x);frontNode(c,x);return;}CacheNode*x=malloc(sizeof(CacheNode));x->key=key;x->val=val;frontNode(c,x);if(++c->size>c->cap){CacheNode*z=c->tail->prev;unlinkNode(c,z);free(z);c->size--;}}
/* Q85 */
/* Two-heap implementation using sorted dynamic arrays for portability. */
typedef struct{long long*a;int n,cap;}Arr;void q85_add(Arr*h,long long x){if(h->n==h->cap){h->cap=h->cap?2*h->cap:8;h->a=realloc(h->a,h->cap*sizeof(long long));}h->a[h->n++]=x;}double q85_median(Arr*h){for(int i=1;i<h->n;i++){long long x=h->a[i];int j=i-1;while(j>=0&&h->a[j]>x)h->a[j+1]=h->a[j],j--;h->a[j+1]=x;}return h->n%2?h->a[h->n/2]:(h->a[h->n/2-1]+h->a[h->n/2])/2.0;}
/* Q86 */
char*q86_serialize(TreeNode*r){if(!r){char*x=strdup("#");return x;}char*a=q86_serialize(r->left),*b=q86_serialize(r->right);int n=snprintf(NULL,0,"%d,%s,%s",r->val,a,b);char*x=malloc(n+1);sprintf(x,"%d,%s,%s",r->val,a,b);free(a);free(b);return x;}
/* Q87 */
int q87_max_path_sum(TreeNode*r){int ans=INT_MIN;int f(TreeNode*x){if(!x)return 0;int a=f(x->left),b=f(x->right);if(a<0)a=0;if(b<0)b=0;int v=x->val+a+b;if(v>ans)ans=v;return x->val+(a>b?a:b);}f(r);return ans;}
/* Q88 */
void q88_recover_bst(TreeNode*r){TreeNode*st[4096],*first=NULL,*second=NULL,*prev=NULL;int top=0;while(r||top){while(r)st[top++]=r,r=r->left;r=st[--top];if(prev&&prev->val>r->val){if(!first)first=prev;second=r;}prev=r;r=r->right;}if(first){int t=first->val;first->val=second->val;second->val=t;}}
/* Q89 */
typedef struct TrieNode{int end;struct TrieNode*c[26];}TrieNode;TrieNode*q89_new(){return calloc(1,sizeof(TrieNode));}void q89_insert(TrieNode*r,const char*s){for(;*s;s++){int i=*s-'a';if(!r->c[i])r->c[i]=q89_new();r=r->c[i];}r->end=1;}bool q89_search(TrieNode*r,const char*s){for(;*s;s++){r=r->c[*s-'a'];if(!r)return false;}return r->end;}bool q89_starts_with(TrieNode*r,const char*s){for(;*s;s++){r=r->c[*s-'a'];if(!r)return false;}return true;}
/* Q90 */
void q90_word_search_ii_dfs(char**b,int n,int m,int i,int j,TrieNode*p,char*path,int d,char**out,int*outn){char ch=b[i][j];if(ch=='#'||!p->c[ch-'a'])return;p=p->c[ch-'a'];path[d]=ch;path[d+1]=0;if(p->end){out[*outn]=strdup(path);p->end=0;(*outn)++;}b[i][j]='#';int dx[4]={1,-1,0,0},dy[4]={0,0,1,-1};for(int k=0;k<4;k++){int x=i+dx[k],y=j+dy[k];if(x>=0&&x<n&&y>=0&&y<m)q90_word_search_ii_dfs(b,n,m,x,y,p,path,d+1,out,outn);}b[i][j]=ch;}
/* Q91 */
long long*q91_dijkstra(int n,int edges[][3],int e,int src){long long*dist=malloc(n*sizeof(long long));char*used=calloc(n,1);for(int i=0;i<n;i++)dist[i]=LLONG_MAX/4;dist[src]=0;for(int it=0;it<n;it++){int u=-1;for(int i=0;i<n;i++)if(!used[i]&&(u<0||dist[i]<dist[u]))u=i;if(u<0||dist[u]>=LLONG_MAX/4)break;used[u]=1;for(int z=0;z<e;z++)if(edges[z][0]==u&&dist[edges[z][1]]>dist[u]+edges[z][2])dist[edges[z][1]]=dist[u]+edges[z][2];}free(used);return dist;}
int q92_network_delay(int n,int e[][3],int m,int src){long long*d=q91_dijkstra(n,e,m,src-1);long long mx=0;for(int i=0;i<n;i++){if(d[i]>=LLONG_MAX/4){free(d);return -1;}if(d[i]>mx)mx=d[i];}free(d);return (int)mx;}
typedef struct{int*p,*r;}DSU;DSU dsu_new(int n){DSU d={malloc(n*sizeof(int)),calloc(n,sizeof(int))};for(int i=0;i<n;i++)d.p[i]=i;return d;}int dsu_f(DSU*d,int x){return d->p[x]==x?x:(d->p[x]=dsu_f(d,d->p[x]));}bool dsu_u(DSU*d,int a,int b){a=dsu_f(d,a);b=dsu_f(d,b);if(a==b)return false;if(d->r[a]<d->r[b]){int t=a;a=b;b=t;}d->p[b]=a;if(d->r[a]==d->r[b])d->r[a]++;return true;}
int q93_mst(int n,int e[][3],int m){for(int i=0;i<m;i++)for(int j=i+1;j<m;j++)if(e[j][2]<e[i][2])for(int k=0;k<3;k++){int t=e[i][k];e[i][k]=e[j][k];e[j][k]=t;}DSU d=dsu_new(n);int ans=0;for(int i=0;i<m;i++)if(dsu_u(&d,e[i][0],e[i][1]))ans+=e[i][2];free(d.p);free(d.r);return ans;}
/* Q94 */
void q94_dfs(int u,int pe,int n,int e[][2],int m,int*tin,int*low,int*timer,int*isBridge){tin[u]=low[u]=(*timer)++;for(int i=0;i<m;i++){int v=-1;if(i==pe)continue;if(e[i][0]==u)v=e[i][1];else if(e[i][1]==u)v=e[i][0];else continue;if(tin[v]>=0){if(tin[v]<low[u])low[u]=tin[v];}else{q94_dfs(v,i,n,e,m,tin,low,timer,isBridge);if(low[v]<low[u])low[u]=low[v];if(low[v]>tin[u])isBridge[i]=1;}}}void q94_critical_connections(int n,int e[][2],int m,int*out){int*tin=malloc(n*sizeof(int)),*low=malloc(n*sizeof(int));for(int i=0;i<n;i++)tin[i]=-1;memset(out,0,m*sizeof(int));int timer=0;for(int i=0;i<n;i++)if(tin[i]<0)q94_dfs(i,-1,n,e,m,tin,low,&timer,out);free(tin);free(low);}
/* Q95 */
void q95_scc(int n,int e[][2],int m,int*comp){/* Kosaraju with O(n^2) adjacency for clarity. */int*vis=calloc(n,1),*ord=malloc(n*sizeof(int)),k=0;void f(int u){vis[u]=1;for(int i=0;i<m;i++)if(e[i][0]==u&&!vis[e[i][1]])f(e[i][1]);ord[k++]=u;}for(int i=0;i<n;i++)if(!vis[i])f(i);for(int i=0;i<n;i++)comp[i]=-1;void r(int u,int c){comp[u]=c;for(int i=0;i<m;i++)if(e[i][1]==u&&comp[e[i][0]]<0)r(e[i][0],c);}int c=0;for(int i=n-1;i>=0;i--)if(comp[ord[i]]<0)r(ord[i],c++);free(vis);free(ord);}
/* Q96 */
char*q96_alien_dictionary(const char**w,int n){int edge[26][26]={0},deg[26]={0},present[26]={0};for(int i=0;i<n;i++)for(int j=0;w[i][j];j++)present[w[i][j]-'a']=1;for(int i=1;i<n;i++){int k=0;while(w[i-1][k]&&w[i][k]&&w[i-1][k]==w[i][k])k++;if(!w[i][k]&&!w[i-1][k])continue;if(!w[i][k]&&w[i-1][k])return strdup("");if(w[i-1][k]&&w[i][k]&&!edge[w[i-1][k]-'a'][w[i][k]-'a'])edge[w[i-1][k]-'a'][w[i][k]-'a']=1,deg[w[i][k]-'a']++;}char*q=malloc(27);int h=0,t=0;char z[26];for(int i=0;i<26;i++)if(present[i]&&!deg[i])z[t++]=i;while(h<t){int u=z[h++];q[h-1]=(char)('a'+u);for(int v=0;v<26;v++)if(edge[u][v]&&!--deg[v])z[t++]=v;}q[t]=0;int cnt=0;for(int i=0;i<26;i++)cnt+=present[i];if(t!=cnt)q[0]=0;return q;}
bool q97_partition_equal(int*a,int n){int sum=0;for(int i=0;i<n;i++)sum+=a[i];if(sum%2)return false;int t=sum/2;char*dp=calloc(t+1,1);dp[0]=1;for(int i=0;i<n;i++)for(int s=t;s>=a[i];s--)dp[s]|=dp[s-a[i]];bool ans=dp[t];free(dp);return ans;}
int q98_min_cut(const char*s){int n=strlen(s);int*cut=malloc(n*sizeof(int));for(int i=0;i<n;i++)cut[i]=i;for(int r=0;r<n;r++){for(int l=r;l>=0&&s[l]==s[r];l--)cut[r]=l? (cut[r]<cut[l-1]+1?cut[r]:cut[l-1]+1):0;for(int l=r-1;l>=0&&s[l]==s[r];l--)cut[r]=l? (cut[r]<cut[l-1]+1?cut[r]:cut[l-1]+1):0;}int ans=n?cut[n-1]:0;free(cut);return ans;}
long long q99_nqueens(int n){long long ans=0;void f(int r,int cols,int d1,int d2){if(r==n){ans++;return;}int avail=((1<<n)-1)&~(cols|d1|d2);while(avail){int b=avail&-avail;avail-=b;f(r+1,cols|b,(d1|b)<<1,(d2|b)>>1);}}f(0,0,0,0);return ans;}
int q100_burst_balloons(int*a,int n){
 int N=n+2; int*b=malloc(N*sizeof(int)); b[0]=b[N-1]=1; for(int i=0;i<n;i++)b[i+1]=a[i];
 int*dp=calloc(N*N,sizeof(int));
 #define DP(i,j) dp[(i)*N+(j)]
 for(int len=2;len<N;len++) for(int l=0;l+len<N;l++){int r=l+len; for(int k=l+1;k<r;k++){int v=DP(l,k)+b[l]*b[k]*b[r]+DP(k,r); if(v>DP(l,r))DP(l,r)=v;}}
 int ans=DP(0,N-1); free(b); free(dp); return ans;
}

File: DSA/Hard/Python/solutions.py
# DSA 100 - Hard | Python solutions
from collections import Counter,defaultdict,deque
import heapq

def q71_trapping_rain_water(h):
 l,r=0,len(h)-1;lm=rm=ans=0
 while l<=r:
  if h[l]<=h[r]:lm=max(lm,h[l]);ans+=lm-h[l];l+=1
  else:rm=max(rm,h[r]);ans+=rm-h[r];r-=1
 return ans
def q72_first_missing_positive(a):
 n=len(a)
 for i,x in enumerate(a):a[i]=x if 1<=x<=n else n+1
 for x in a:
  v=abs(x)
  if v<=n and a[v-1]>0:a[v-1]=-a[v-1]
 return next((i+1 for i,x in enumerate(a) if x>0),n+1)
def q73_median_two_sorted(a,b):
 if len(a)>len(b):a,b=b,a
 m,n=len(a),len(b);lo,hi=0,m
 while lo<=hi:
  i=(lo+hi)//2;j=(m+n+1)//2-i
  al=a[i-1] if i else float('-inf');ar=a[i] if i<m else float('inf')
  bl=b[j-1] if j else float('-inf');br=b[j] if j<n else float('inf')
  if al<=br and bl<=ar:return max(al,bl) if (m+n)%2 else (max(al,bl)+min(ar,br))/2
  if al>br:hi=i-1
  else:lo=i+1
def q74_largest_rectangle(h):
 st=[];ans=0
 for i,x in enumerate(h+[0]):
  while st and h[st[-1]]>x:
   j=st.pop();left=st[-1] if st else -1;ans=max(ans,h[j]*(i-left-1))
  st.append(i)
 return ans
def q75_max_product(a):
 mx=mn=ans=a[0]
 for x in a[1:]:
  if x<0:mx,mn=mn,mx
  mx=max(x,mx*x);mn=min(x,mn*x);ans=max(ans,mx)
 return ans
def q76_count_inversions(a):
 def merge(a):
  if len(a)<2:return a,0
  m=len(a)//2;l,x=merge(a[:m]);r,y=merge(a[m:]);i=j=inv=0;res=[]
  while i<len(l) and j<len(r):
   if l[i]<=r[j]:res.append(l[i]);i+=1
   else:res.append(r[j]);j+=1;inv+=len(l)-i
  return res+l[i:]+r[j:],x+y+inv
 _,ans=merge(a);return ans
def q77_min_window(s,t):
 need=Counter(t);have=Counter();missing=len(t);l=st=0;best=float('inf')
 for r,c in enumerate(s):
  have[c]+=1
  if have[c]<=need[c]:missing-=1
  while missing==0:
   if r-l+1<best:best=r-l+1;st=l
   have[s[l]]-=1
   if have[s[l]]<need[s[l]]:missing+=1
   l+=1
 return '' if best==float('inf') else s[st:st+best]
def q78_word_break(s,words):
 d=set(words);p={0:None}
 for i in range(len(s)+1):
  if i not in p:continue
  for j in range(i+1,len(s)+1):
   if s[i:j] in d:p[j]=(i,s[i:j])
 if len(s) not in p:return []
 out=[];i=len(s)
 while i: i0,w=p[i];out.append(w);i=i0
 return out[::-1]
def q79_regex_match(s,p):
 from functools import lru_cache
 @lru_cache(None)
 def f(i,j):
  if j==len(p):return i==len(s)
  first=i<len(s) and (p[j]==s[i] or p[j]=='.')
  if j+1<len(p) and p[j+1]=='*':return f(i,j+2) or (first and f(i+1,j))
  return first and f(i+1,j+1)
 return f(0,0)
def q80_edit_distance(a,b):
 dp=list(range(len(b)+1))
 for i,x in enumerate(a,1):
  nd=[i]+[0]*len(b)
  for j,y in enumerate(b,1):nd[j]=min(dp[j]+1,nd[j-1]+1,dp[j-1]+(x!=y))
  dp=nd
 return dp[-1]
def q81_longest_pal_subseq(s):
 n=len(s);dp=[0]*n
 for i in range(n-1,-1,-1):
  prev=0;dp[i]=1
  for j in range(i+1,n):
   old=dp[j];dp[j]=prev+2 if s[i]==s[j] else max(dp[j],dp[j-1]);prev=old
 return dp[-1] if s else 0
class Node:
 def __init__(self,val,next=None):self.val=val;self.next=next
def q82_merge_k_lists(lists):
 h=[];uid=0
 for x in lists:
  if x:heapq.heappush(h,(x.val,uid,x));uid+=1
 d=Node(0);t=d
 while h:
  _,_,x=heapq.heappop(h);t.next=x;t=x
  if x.next:heapq.heappush(h,(x.next.val,uid,x.next));uid+=1
 return d.next
def q83_reverse_k_group(h,k):
 d=Node(0,h);g=d
 while True:
  p=g
  for _ in range(k):p=p.next if p else None
  if not p:break
  cur=g.next;prev=p.next
  for _ in range(k):n=cur.next;cur.next=prev;prev=cur;cur=n
  old=g.next;g.next=prev;g=old
 return d.next
class q84_LRUCache:
 def __init__(self,capacity):self.c=capacity;self.d={};self.order=[]
 def get(self,key):
  if key not in self.d:return -1
  self.order.remove(key);self.order.append(key);return self.d[key]
 def put(self,key,value):
  if key in self.d:self.order.remove(key)
  self.d[key]=value;self.order.append(key)
  if len(self.d)>self.c:x=self.order.pop(0);del self.d[x]
class q85_MedianFinder:
 def __init__(self):self.lo=[];self.hi=[]
 def addNum(self,x):
  heapq.heappush(self.lo,-x);heapq.heappush(self.hi,-heapq.heappop(self.lo))
  if len(self.hi)>len(self.lo):heapq.heappush(self.lo,-heapq.heappop(self.hi))
 def findMedian(self):
  return -self.lo[0] if len(self.lo)>len(self.hi) else (-self.lo[0]+self.hi[0])/2
class TNode:
 def __init__(self,val,left=None,right=None):self.val=val;self.left=left;self.right=right
def q86_serialize(r):
 out=[]
 def f(x):
  if not x:out.append('#');return
  out.append(str(x.val));f(x.left);f(x.right)
 f(r);return ','.join(out)
def q86_deserialize(s):
 it=iter(s.split(','))
 def f():
  x=next(it)
  if x=='#':return None
  r=TNode(int(x));r.left=f();r.right=f();return r
 return f()
def q87_max_path_sum(r):
 ans=[float('-inf')]
 def f(x):
  if not x:return 0
  a=max(0,f(x.left));b=max(0,f(x.right));ans[0]=max(ans[0],x.val+a+b);return x.val+max(a,b)
 f(r);return ans[0]
def q88_recover_bst(r):
 st=[];prev=first=second=None;x=r
 while st or x:
  while x:st.append(x);x=x.left
  x=st.pop()
  if prev and prev.val>x.val:
   if first is None:first=prev
   second=x
  prev=x;x=x.right
 first.val,second.val=second.val,first.val
class q89_Trie:
 def __init__(self):self.root={}
 def insert(self,w):
  p=self.root
  for c in w:p=p.setdefault(c,{})
  p['#']=True
 def search(self,w):
  p=self.root
  for c in w:
   if c not in p:return False
   p=p[c]
  return '#' in p
 def startsWith(self,w):
  p=self.root
  for c in w:
   if c not in p:return False
   p=p[c]
  return True
def q90_word_search_ii(board,words):
 root={}
 for w in words:
  p=root
  for c in w:p=p.setdefault(c,{})
  p['#']=w
 out=set();n=len(board);m=len(board[0])
 def dfs(i,j,p):
  c=board[i][j]
  if c=='#' or c not in p:return
  q=p[c]
  if '#' in q:out.add(q['#'])
  board[i][j]='#'
  for di,dj in ((1,0),(-1,0),(0,1),(0,-1)):
   x,y=i+di,j+dj
   if 0<=x<n and 0<=y<m:dfs(x,y,q)
  board[i][j]=c
 for i in range(n):
  for j in range(m):dfs(i,j,root)
 return list(out)
def q91_dijkstra(n,edges,src):
 g=[[] for _ in range(n)]
 for u,v,w in edges:g[u].append((v,w))
 d=[float('inf')]*n;d[src]=0;h=[(0,src)]
 while h:
  du,u=heapq.heappop(h)
  if du!=d[u]:continue
  for v,w in g[u]:
   if du+w<d[v]:d[v]=du+w;heapq.heappush(h,(d[v],v))
 return d
def q92_network_delay(n,times,k):
 d=q91_dijkstra(n,[(u-1,v-1,w) for u,v,w in times],k-1);x=max(d);return -1 if x==float('inf') else x
def q93_mst(n,edges):
 p=list(range(n));
 def f(x):
  while p[x]!=x:p[x]=p[p[x]];x=p[x]
  return x
 ans=0
 for u,v,w in sorted(edges,key=lambda e:e[2]):
  a,b=f(u),f(v)
  if a!=b:p[a]=b;ans+=w
 return ans
def q94_critical_connections(n,edges):
 g=[[] for _ in range(n)];
 for i,(u,v) in enumerate(edges):g[u].append((v,i));g[v].append((u,i))
 tin=[-1]*n;low=[0]*n;ans=[];time=0
 def dfs(u,pe):
  nonlocal time;tin[u]=low[u]=time;time+=1
  for v,eid in g[u]:
   if eid==pe:continue
   if tin[v]>=0:low[u]=min(low[u],tin[v])
   else:
    dfs(v,eid);low[u]=min(low[u],low[v])
    if low[v]>tin[u]:ans.append([u,v])
 dfs(0,-1);return ans
def q95_scc(n,edges):
 g=[[] for _ in range(n)];rg=[[] for _ in range(n)]
 for u,v in edges:g[u].append(v);rg[v].append(u)
 vis=[0]*n;order=[]
 def f(u):
  vis[u]=1
  for v in g[u]:
   if not vis[v]:f(v)
  order.append(u)
 for i in range(n):
  if not vis[i]:f(i)
 comp=[-1]*n;cid=0
 def r(u):
  comp[u]=cid
  for v in rg[u]:
   if comp[v]<0:r(v)
 for u in reversed(order):
  if comp[u]<0:r(u);cid+=1
 return comp
def q96_alien_dictionary(words):
 chars=set(''.join(words));g={c:set() for c in chars};d={c:0 for c in chars}
 for a,b in zip(words,words[1:]):
  k=0
  while k<len(a) and k<len(b) and a[k]==b[k]:k+=1
  if k==len(b)<len(a):return ''
  if k<len(a) and k<len(b) and b[k] not in g[a[k]]:g[a[k]].add(b[k]);d[b[k]]+=1
 q=deque(c for c in chars if d[c]==0);out=''
 while q:
  c=q.popleft();out+=c
  for v in g[c]:d[v]-=1;q.append(v) if d[v]==0 else None
 return out if len(out)==len(chars) else ''
def q97_partition_equal(a):
 s=sum(a)
 if s%2:return False
 t=s//2;dp={0}
 for x in a:dp|={v+x for v in list(dp) if v+x<=t}
 return t in dp
def q98_palindrome_partition_min_cuts(s):
 n=len(s);pal=[[False]*n for _ in range(n)];cut=list(range(n))
 for r in range(n):
  for l in range(r+1):
   if s[l]==s[r] and (r-l<2 or pal[l+1][r-1]):
    pal[l][r]=True;cut[r]=0 if l==0 else min(cut[r],cut[l-1]+1)
 return cut[-1] if n else 0
def q99_n_queens(n):
 ans=[]
 def dfs(r,cols,d1,d2):
  if r==n:ans.append(1);return
  for c in range(n):
   if c not in cols and r+c not in d1 and r-c not in d2:dfs(r+1,cols|{c},d1|{r+c},d2|{r-c})
 dfs(0,set(),set(),set());return len(ans)
def q100_burst_balloons(a):
 a=[1]+a+[1];n=len(a);dp=[[0]*n for _ in range(n)]
 for length in range(2,n):
  for l in range(n-length):
   r=l+length
   dp[l][r]=max((dp[l][k]+a[l]*a[k]*a[r]+dp[k][r] for k in range(l+1,r)),default=0)
 return dp[0][-1]
